h2o-core/src/main/java/water/rapids/ast/prims/operators/AstBinOp.java

```diff
@@ -329,15 +343,47 @@ public void map(Chunk[] chks, NewChunk[] cress) {
           if (clf.vec().isString())
             for (int i = 0; i < clf._len; i++)
               cres.addNum(str_op(clf.atStr(lfstr, i), crt.atStr(rtstr, i)));
-          else
+          else if (categoricalOK() && clf.vec().isCategorical() && crt.vec().isCategorical()) {
+            // The vec with longer domain is iterated over due to categorical mapping
+            if (clf.vec().domain().length > crt.vec().domain().length) {
+              for (int i = 0; i < clf._len; i++) {
+                cres.addNum(op(clf.atd(i), alignedCategoricals[c][(int) crt.atd(i)]));
```

Comment:
I think this will fail for NAs in the data. Casting Double.NaN to int will not work. NAs should be passed without transformation.

Reply:
Indeed. Casting Double.NaN to 32-bit int results in the value of 0.